python/tvm/relax/frontend/torch/dynamo.py

```diff
@@ -0,0 +1,156 @@
+# mypy: ignore-errors
+"""PyTorch Dynamo backend of Relax."""
+import functools
+from typing import Optional
+
+import tvm
+from tvm.relax.vm import build as relax_build
+from tvm.relax.frontend.torch.fx_translator import from_fx
+
+
+def device_from_inputs(example_inputs):
+    for x in example_inputs:
+        if hasattr(x, "device"):
+            return x.device
+    return None
+
+
+def relax_dynamo(pipeline: Optional[tvm.transform.Pass] = None):
+    """A helper function to create a relax backend.
+
+    Parameters
+    ----------
+    pipeline : Optional[tvm.transform.Pass]
+        The pipeline to be applied to the relax module before sent to build.
+
+    Returns
+    -------
+    backend : Callable[[torch.fx.GraphModule, List[torch.Tensor]], Callable]
+        The relax dynamo backend.
+    """
+
+    def _relax_backend(graph_module, example_inputs):
+        import torch  # type: ignore[import]
+
+        assert isinstance(graph_module, torch.fx.GraphModule)
+
+        def to_torch_tensor(nd_tensor):
+            """A helper function to transfer a NDArray to torch.tensor."""
+            if isinstance(nd_tensor, tvm.nd.NDArray):
+                return torch.from_numpy(nd_tensor.numpy())
+            elif isinstance(nd_tensor, tvm.ir.Array):
+                return tuple(to_torch_tensor(x) for x in nd_tensor)
+            else:
+                raise ValueError(f"Unsupported type {type(nd_tensor)}")
+
+        def to_tvm_tensor(torch_tensor):
+            """A helper function to transfer a torch.tensor to NDArray."""
+            if not isinstance(torch_tensor, torch._subclasses.fake_tensor.FakeTensor):
+                return tvm.nd.array(torch_tensor.numpy())
+            # Fake Tensor
+            real_tensor = torch.randn(torch_tensor.shape, dtype=torch_tensor.dtype)
+            return tvm.nd.array(real_tensor.numpy())
+
+        device = device_from_inputs(example_inputs)
+        input_info = [(tuple(tensor.shape), str(tensor.dtype)) for tensor in example_inputs]
+        mod = from_fx(graph_module, input_info)
+
+        if device.type == "cuda":
+            dev = tvm.cuda(device.index)
+            target = tvm.target.cuda()
+        else:
+            dev = tvm.cpu(0)
+            target = tvm.target.Target(llvm_target())
+
+        # invoke optimization pipeline.
+        if pipeline is None:
+            # get default pipeline
+            seq = tvm.relax.get_pipeline()
+        elif isinstance(pipeline, str):
+            # lookup by name
+            seq = tvm.relax.get_pipeline(pipeline)
+        else:
+            seq = pipeline
+
+        mod = mod.with_attr("target", target)
+        mod = seq(mod)
+
+        ex = relax_build(mod, target=target)
+
+        vm = tvm.relax.vm.VirtualMachine(exec=ex.mod, device=dev)
+
+        def exec_tvm(*i_args):
+            args = [a.contiguous() for a in i_args]
+            vm_args = list()
+            for arg in args:
+                if arg.dim() != 0:
```

Comment:
Hi. I'm confused about this line. Actually, I'm working with a model that takes a float value as one of its inputs. So why was the value with no dim ignored? Thanks!

Reply:
Hi @liquanfeng, likely this is a redundant check. You're welcome to contribute and fix it once confirmed.